File: CoolOptions.hs
module CoolOptions where

data CoolOptions = CoolOptions
  { scanDebug            :: Bool
  , parseDebug           :: Bool
  , semantDebug          :: Bool
  , analysisDebug        :: Bool
  , cgenDebug            :: Bool
  , scan                 :: Bool
  , parse                :: Bool
  , semant               :: Bool
  , analyze              :: Bool
  , optimize             :: Bool
  , codegen              :: Bool
  , enableGc             :: Bool
  , rightToLeftChildren  :: Bool
  , genC                 :: Bool
  , basicFile            :: String
  , analysis             :: String
  , outFilename          :: Maybe String
  , maxErrors            :: Int
  , stripAttributes      :: Int
  } deriving (Show, Eq)

defaultCoolOptions :: CoolOptions
defaultCoolOptions = CoolOptions
  { scanDebug = False
  , parseDebug = False
  , semantDebug = False
  , analysisDebug = False
  , cgenDebug = False
  , scan = False
  , parse = True
  , semant = True
  , analyze = False
  , optimize = False
  , codegen = True
  , enableGc = False
  , rightToLeftChildren = False
  , genC = False
  , basicFile = "/usr/local/lib/basic.cool"
  , analysis = "resolve:inline:dead-let"
  , outFilename = Nothing
  , maxErrors = 50
  , stripAttributes = 0
  }

-- Getters
getScanDebug :: CoolOptions -> Bool
getScanDebug = scanDebug

getParseDebug :: CoolOptions -> Bool
getParseDebug = parseDebug

getSemantDebug :: CoolOptions -> Bool
getSemantDebug = semantDebug

getAnalysisDebug :: CoolOptions -> Bool
getAnalysisDebug = analysisDebug

getCgenDebug :: CoolOptions -> Bool
getCgenDebug = cgenDebug

getScan :: CoolOptions -> Bool
getScan = scan

getParse :: CoolOptions -> Bool
getParse = parse

getSemant :: CoolOptions -> Bool
getSemant = semant

getAnalyze :: CoolOptions -> Bool
getAnalyze = analyze

getOptimize :: CoolOptions -> Bool
getOptimize = optimize

getCodegen :: CoolOptions -> Bool
getCodegen = codegen

getEnableGc :: CoolOptions -> Bool
getEnableGc = enableGc

getRightToLeftChildren :: CoolOptions -> Bool
getRightToLeftChildren = rightToLeftChildren

getGenC :: CoolOptions -> Bool
getGenC = genC

getBasicFile :: CoolOptions -> String
getBasicFile = basicFile

getAnalysis :: CoolOptions -> String
getAnalysis = analysis

getOutFilename :: CoolOptions -> Maybe String
getOutFilename = outFilename

getMaxErrors :: CoolOptions -> Int
getMaxErrors = maxErrors

getStripAttributes :: CoolOptions -> Int
getStripAttributes = stripAttributes

setScanDebug :: Bool -> CoolOptions -> CoolOptions
setScanDebug val opts = opts { scanDebug = val }

setParseDebug :: Bool -> CoolOptions -> CoolOptions
setParseDebug val opts = opts { parseDebug = val }

setSemantDebug :: Bool -> CoolOptions -> CoolOptions
setSemantDebug val opts = opts { semantDebug = val }

setAnalysisDebug :: Bool -> CoolOptions -> CoolOptions
setAnalysisDebug val opts = opts { analysisDebug = val }

setCgenDebug :: Bool -> CoolOptions -> CoolOptions
setCgenDebug val opts = opts { cgenDebug = val }

setScan :: Bool -> CoolOptions -> CoolOptions
setScan val opts = opts { scan = val }

setParse :: Bool -> CoolOptions -> CoolOptions
setParse val opts = opts { parse = val }

setSemant :: Bool -> CoolOptions -> CoolOptions
setSemant val opts = opts { semant = val }

setAnalyze :: Bool -> CoolOptions -> CoolOptions
setAnalyze val opts = opts { analyze = val }

setOptimize :: Bool -> CoolOptions -> CoolOptions
setOptimize val opts = opts { optimize = val }

setCodegen :: Bool -> CoolOptions -> CoolOptions
setCodegen val opts = opts { codegen = val }

setEnableGc :: Bool -> CoolOptions -> CoolOptions
setEnableGc val opts = opts { enableGc = val }

setRightToLeftChildren :: Bool -> CoolOptions -> CoolOptions
setRightToLeftChildren val opts = opts { rightToLeftChildren = val }

setGenC :: Bool -> CoolOptions -> CoolOptions
setGenC val opts = opts { genC = val }

setBasicFile :: String -> CoolOptions -> CoolOptions
setBasicFile val opts = opts { basicFile = val }

setAnalysis :: String -> CoolOptions -> CoolOptions
setAnalysis val opts = opts { analysis = val }

setOutFilename :: Maybe String -> CoolOptions -> CoolOptions
setOutFilename val opts = opts { outFilename = val }

setMaxErrors :: Int -> CoolOptions -> CoolOptions
setMaxErrors val opts = opts { maxErrors = val }

setStripAttributes :: Int -> CoolOptions -> CoolOptions
setStripAttributes val opts = opts { stripAttributes = val }

File: CoolOptions.scala
class CoolOptions()
{
  // Boolean flags

  var scan_debug : Boolean = false;
  def get_scan_debug() : Boolean = scan_debug;
  def set_scan_debug(f : Boolean) : Unit = scan_debug = f;

  var parse_debug : Boolean = false;
  def get_parse_debug() : Boolean = parse_debug;
  def set_parse_debug(f : Boolean) : Unit = parse_debug = f;

  var semant_debug : Boolean = false;
  def get_semant_debug() : Boolean = semant_debug;
  def set_semant_debug(f : Boolean) : Unit = semant_debug = f;

  var analysis_debug : Boolean = false;
  def get_analysis_debug() : Boolean = analysis_debug;
  def set_analysis_debug(f : Boolean) : Unit = analysis_debug = f;

  var cgen_debug : Boolean = false;
  def get_cgen_debug() : Boolean = cgen_debug;
  def set_cgen_debug(f : Boolean) : Unit = cgen_debug = f;

  var scan : Boolean = false;
  def get_scan() : Boolean = scan;
  def set_scan(f : Boolean) : Unit = scan = f;

  var parse : Boolean = true;
  def get_parse() : Boolean = parse;
  def set_parse(f : Boolean) : Unit = parse = f;

  var semant : Boolean = true;
  def get_semant() : Boolean = semant;
  def set_semant(f : Boolean) : Unit = semant = f;

  var analyze : Boolean = false;
  def get_analyze() : Boolean = analyze;
  def set_analyze(f : Boolean) : Unit = analyze = f;

  var optimize : Boolean = false;
  def get_optimize() : Boolean = optimize;
  def set_optimize(f : Boolean) : Unit = optimize = f;

  var codegen : Boolean = true;
  def get_codegen() : Boolean = codegen;
  def set_codegen(f : Boolean) : Unit = codegen = f;

  var enable_gc : Boolean = false;
  def get_enable_gc() : Boolean = enable_gc;
  def set_enable_gc(f : Boolean) : Unit = enable_gc = f;

  var right_to_left_children : Boolean = false;
  def get_right_to_left_children() : Boolean = right_to_left_children;
  def set_right_to_left_children(f : Boolean) : Unit = right_to_left_children = f;

  var gen_c : Boolean = false;
  def get_gen_c() : Boolean = gen_c;
  def set_gen_c(f : Boolean) : Unit = gen_c = f;


  // String flags

  var basic_file:String = "/usr/local/lib/basic.cool";
  def get_basic_file() : String = basic_file;
  def set_basic_file(f : String) : Unit = basic_file = f;

  var analysis : String = "resolve:inline:dead-let";
  def get_analysis() : String = analysis;
  def set_analysis(f : String) : Unit = analysis = f;

  var out_filename : String = null;
  def get_out_filename() : String = out_filename;
  def set_out_filename(f : String) : Unit = out_filename = f;


  // Int flags

  var max_errors : Int = 50;
  def get_max_errors() : Int = max_errors;
  def set_max_errors(f : Int) : Unit = max_errors = f;

  var strip_attributes : Int = 0;
  def get_strip_attributes() : Int = strip_attributes;
  def set_strip_attributes(f : Int) : Unit = strip_attributes = f;
}
